File: python/medium/Convert_Sorted_List_to_Binary_Search_Tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
        if not head:
            return None
        num_list = []
        while head.next != None:
            num_list.append(head.val)
            head = head.next
        num_list.append(head.val)
        def next_node(node, list_n):
            pivot = int(len(list_n)/2)
            pre = list_n[:pivot]
            back = list_n[pivot+1:]
            node.val = list_n[pivot]
            if len(pre) > 0:
                node.left = TreeNode()
                next_node(node.left, pre)
            if len(back) > 0:
                node.right = TreeNode()
                next_node(node.right, back)
        pivot = int(len(num_list)/2)
        pre = num_list[:pivot]
        back = num_list[pivot+1:]
        tree = TreeNode(num_list[pivot])
        if len(pre) > 0:
            tree.left = TreeNode()
            next_node(tree.left, pre)
        if len(back) > 0:
            tree.right = TreeNode()
            next_node(tree.right, back)
        return tree

File: cpp/medium/Convert_Sorted_List_to_Binary_Search_Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        if(head == nullptr) return nullptr;

        vector<TreeNode*> node_list;
        TreeNode* node_head;
        TreeNode* node;
        node_head = new TreeNode(head->val);
        head = head->next;
        int cnt = 1;
        while (head != nullptr) {
            node = new TreeNode(head->val);
            node->left = node_head;
            node_head = node;
            head = head->next;
            cnt++;
        }

        if (cnt <= 2) return node_head;
        TreeNode* root = divide_conquer(node_head, cnt);

        return root;
    }
    TreeNode* divide_conquer(TreeNode* node_head, int cnt) {
        if (cnt <= 2) return node_head;
        TreeNode* right_root = node_head;
        int mid = cnt / 2;
        if (mid % 2 == 0) mid--;
        for (int i = 0; i < mid - 1; ++i) {
            node_head = node_head->left;
        }
        TreeNode* root = node_head->left;
        node_head->left = nullptr;
        root->right = divide_conquer(right_root, mid);
        root->left = divide_conquer(root->left, cnt - mid - 1);
        return root;
    }
};
